beamer/check/commands.py

```diff
@@ -264,3 +264,166 @@ def initiate_l1_invalidations(
+    password: str,
+    rpc_file: Path,
+    artifacts_dir: Path,
+    abi_dir: Path,
+    file: Path,
+) -> None:
+    """Verify L1 invalidations stored in FILE."""
+    beamer.util.setup_logging(log_level="DEBUG", log_json=False)
+
+    account = beamer.util.account_from_keyfile(keystore_file, password)
+    log.info("Loaded keystore file", address=account.address)
+
+    abi_manager = ABIManager(abi_dir)
+    rpc_info = beamer.util.load_rpc_info(rpc_file)
+
+    with file.open("rt") as f:
+        data = json.load(f)
+    invalidations = apischema.deserialize(list[Invalidation], data)
+
+    # Pairs (proof_source, proof_target) for which an invalidation has already been
+    # verified.
+    verified = set()
+
+    errors = 0
+    chain_pairs = set()
+    for idx, invalidation in enumerate(invalidations):
+        pair = invalidation.proof_source, invalidation.proof_target
+        chain_pairs.add(pair)
+
+        if pair in verified:
+            log.info(
+                "Verified invalidation already exists for chain pair, skipping",
+                idx=idx,
+                proof_source=pair[0],
+                proof_target=pair[1],
+            )
+            continue
+
+        # First check the finalization timestamp.
+        if (
+            invalidation.finalization_timestamp is None
+            or datetime.utcnow() <= invalidation.finalization_timestamp
+        ):
+            log.error(
+                "Invalidation not yet finalized, skipping",
+                idx=idx,
+                proof_source=pair[0],
+                proof_target=pair[1],
+            )
+            errors += 1
+            continue
+
+        deployment = beamer.artifacts.load(artifacts_dir, invalidation.proof_target)
+        url = rpc_info[invalidation.proof_target]
+        w3 = make_web3(url, account)
+        assert w3.eth.chain_id == invalidation.proof_target
+        log.info("Connected to RPC", url=url)
+
+        # Check the request manager directly to see if it has been already verifed.
+        request_manager = obtain_contract(w3, abi_manager, deployment, "RequestManager")
+        if request_manager.functions.isInvalidFill(
+            invalidation.request_id, invalidation.fill_id
+        ).call():
+            log.info(
+                "Invalidation verified successfully",
+                idx=idx,
+                proof_source=pair[0],
+                proof_target=pair[1],
+            )
+            verified.add(pair)
+            continue
+
+        log.debug("Starting relayer", txhash=invalidation.txhash)
+        try:
+            run_relayer_for_tx(
+                l1_rpc=rpc_info[deployment.base.chain_id],
+                l2_relay_from_rpc_url=rpc_info[invalidation.proof_source],
+                l2_relay_to_rpc_url=rpc_info[invalidation.proof_target],
+                account=account,
+                tx_hash=invalidation.txhash,
+            )
+        except RelayerError as exc:
+            log.error(
```

Comment:
I noticed an edge case here: 

In line 367 you check if the invalidation is already resolved, and if so, the respective chain pair is gonna be marked as verified. Thus any further invalidations will not be tried to be resolved, which is fine for now.

However, there is a situation, where the first invalidation for the chain pair fails on the relayer, and if I'm not mistaken, the loop will continue and try the next one. This means that the second invalidation is the one which verified the chain pair. 
If you then run the command again, it would try to resolve the first one again even though the chain pair is already verified by the second invalidation. 
It's just a nitpick.

Reply:
Which leads me to the thinking, what is the intended algorithm for this command? Especially if you run the command more often (either due to a crash or because the file contains multiple chain pairs, with different finalization timestamps). 

I understood the code in a way, that it would always check if it was already verified and if so then don't do anything for this chain pair. Is that correct? 

Then I see the following situations, which are really just nitpicks IMO:

1) The situation above where not the first invalidation for a chain pair resolved correctly but a later one did. Then the code will always invalidate the first one again

2) We cannot really do another invalidation with this command if we wanted to. (probably deleting the resolved invalidation from the file would help). However, I don't have a real scenario at hand where we would want to do this, but for sure for some reason we always do multiple invalidations in order to safe another cycle of finality periods.
A tradeoff could be to introduce a flag, which will not check for existing invalidations and thus requires one invalidation for each chain pair again.

Reply:
> I understood the code in a way, that it would always check if it was already verified and if so then don't do anything for this chain pair. Is that correct?

Yes. Which is why I would not consider your comments here nitpics -- it's a bug that needs to be fixed.
We need to 
- group all invalidations by (source, target) chain pair
- drop the whole list for a particular pair if any were had been verified
- verify invalidations of remaining chain pairs

Reply:
To provide more info on how I am thinking about the command:

> 1. The situation above where not the first invalidation for a chain pair resolved correctly but a later one did. Then the code will always invalidate the first one again

This should not happen. Once an invalidation is done, we should not waste time/money for checking the same chain pair again.

> 2\. We cannot really do another invalidation with this command if we wanted to. (probably deleting the resolved invalidation from the file would help). However, I don't have a real scenario at hand where we would want to do this, but for sure for some reason we always do multiple invalidations in order to safe another cycle of finality periods.
>         A tradeoff could be to introduce a flag, which will not check for existing invalidations and thus requires one invalidation for each chain pair again.

That, or a flag to remove the ones which were already verified (perhaps saving the original in a backup file).